### ISSUE TITLE
twisted.web.template is lacking patterns

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5040]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)| @mithrandi reported|
|-|-|
|Trac ID|trac#5040|
|Type|enhancement|
|Created|2011-04-08 14:46:25Z|
|Branch|https://github.com/twisted/twisted/tree/pattern-surrogate-5040-2|

Patterns are a feature that Nevow templates have. They look something like this:

```
#!xml
<div t:render="foo" xmlns:t="http://twistedmatrix.com/ns/twisted.web.template/0.1">
<p>Not part of the pattern.</p>
<p t:pattern="myPattern">More stuff here.</p>
</div>
```

Then, in your Python code, you can do something like this:
```python
@renderer
def foo(self, req, tag):
    copy = tag.onePattern('myPattern')
```

The tag with the `t:template` attribute will be omitted from the final rendering, but the `onePattern` call in the renderer will return a copy of it. This is usually used to either select between several elements to be included in the final output, or in conjunction with slots, to fill lots of copies of a pattern with different values.

Attachments:

* [nestedData.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4af/4afe98453ca74b91f00881ecefe2df8e96fd27b0/1eee9341cf4bdb4082e9ae82d7ef39cd5f1e4d2a.py) (3006 bytes) - added by indigo on 2011-09-29 16:35:41Z - sample program rendering nested data, for illustrative purposes
* [template-namespaces-5040.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4af/4afe98453ca74b91f00881ecefe2df8e96fd27b0/147be99f534a7c0e0b6e4fe3ba4020e9edd1cdfc.patch) (4089 bytes) - added by indigo on 2011-09-30 16:13:57Z - Implementation of twisted.web.template.Namespace
* [nestedDataWithNamespaces.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4af/4afe98453ca74b91f00881ecefe2df8e96fd27b0/0285a147a72f51ba62e8911660300417e9fa6349.py) (1412 bytes) - added by indigo on 2011-09-30 16:14:52Z - solution to nested data rendering problem using Namespace
* [template-namespaces-5040.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4af/4afe98453ca74b91f00881ecefe2df8e96fd27b0/eb2c62478896872b17a2a6d909b7a0983ef4f7c9.patch) (2677 bytes) - added by indigo on 2011-10-14 19:14:31Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5040 5040
type__enhancement enhancement
reporter__mithrandi mithrandi
priority__normal normal
milestone__ 
branch__branches_pattern_surrogate_5040_2 branches/pattern-surrogate-5040-2
branch_author__exarkun__indigo exarkun, indigo
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1302273985000000 1302273985000000
changetime__1318891843000000 1318891843000000
version__None None
owner__exarkun exarkun
cc__jknight cc__indigo
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

I'm not 100% sure that patterns are a good idea (at least, as they're implemented in Nevow).

I don't think you really _need_ them; you _can_ just load a separate template.  And they complicate the interface between templates and code, in a way that we never actually managed to describe formally, or verify statically.  It's hard to write an error message that explains where a pattern isn't.

I'm not saying that we should not add this feature for sure, but there should be very clear use-cases laid out on this ticket, things that demonstrate the way that patterns might be used, and don't involve the words 'foo' or 'bar' or 'my' anywhere.  The only clear use-case I've seen for patterns is implementing alternating-row coloring, but you can do that in modern browsers with `:nth-child(even|odd)` CSS selectors, with no logic in your template at all.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

> implementing alternating-row coloring

Alternating with three colors

**Comment by twisted-trac** [comment]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)<a name="note_4"></a>|@bitglue commented|
|-|-|

Replying to [exarkun](#note_3):
> > implementing alternating-row coloring
> 
> Alternating with three colors

I hate to dilute my own argument, but that also is possible with CSS. See http://www.w3.org/TR/css3-selectors/#nth-child-pseudo . Spoiler:

tr:nth-child(3n+0) { background: #ccc }
tr:nth-child(3n+1) { background: #ddd }
tr:nth-child(3n+2) { background: #eee }

But I digress.

Replying to [glyph](#note_2):
> I don't think you really need them; you can just load a separate template.

Sure, you can just load another template. By the same argument, we can do without docstrings. I mean, you can just put the documentation in another file. For that matter, we can discard the "and", "or", and "not" operators, because you can just use "nand".

You can, but it's just damn inconvenient. If it's in a separate template, it's in a separate file. Files have names. What do I name that pattern that I use for the rows in that one table on that one page? client-listing-client-assets-odd-rows.html? Let's get rid of lambda too, and generator expressions, and the yield statement, because you can just do those things with named functions.

Looking at a fairly simple site I've developed, I have thirteen patterns in three templates. Three files, or sixteen? report-alternate-form-item-rendering-for-when-values-are-immutable.html? With what, all of a <tr><td><n:slot name='value'/></td></tr>?

Not only would I have to name all these things, but now I can't have a static page which will render as a reasonable mock-up of the live version. I commonly insert lorem ipsum in slots to this end. When my render logic has any conditionals, I can use patterns to place "templates" in the visual context they will be used. Loading the template (with each pattern included) in a browser then allows me to do a quick aesthetic check of all the ways a thing might be rendered. It also means I can pass my template through an HTML validation tool and guarantee each pattern is valid in the context where it will be used.

Patterns also facilitate reuse of render methods. Patterns are essentially parameters to renderers. If the view layer can't pass parameters to the renderer, well, that's pretty limiting. Here's a use case: rendering the currently active navigation element differently from the others.

Because renderers in Nevow can take parameters from both the view (patterns) and model (data, which I'm also missing) they are highly reusable. Looking at my same simple site, my templates call a renderer 17 times. All but five of those are the standard renderers "sequence" or "mapping". Of those five, three are just "if CONDITION: return ctx.tag". The other two are just "return generateFormRenderer(self.formDefinition)". Because I can use the standard renderers over 70% of the time, my pages are really easy to test. It's much easier to assert that a data method returned the correct mapping than to assert that a render method unfolded a nested data structure and filled the correct slots.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

Replying to [indigo](#note_4):

> Because renderers in Nevow can take parameters from both the view (patterns) and model (data, which I'm also missing) they are highly reusable. Looking at my same simple site, my templates call a renderer 17 times. All but five of those are the standard renderers "sequence" or "mapping". Of those five, three are just "if CONDITION: return ctx.tag". The other two are just "return generateFormRenderer(self.formDefinition)". Because I can use the standard renderers over 70% of the time, my pages are really easy to test. It's much easier to assert that a data method returned the correct mapping than to assert that a render method unfolded a nested data structure and filled the correct slots.

I've actually found the opposite.  Because renderers can take parameters from the view and the model, determined by the template, the template becomes a functional component of the system which is hard to test.  The difficulty I had was due in part to poor documentation, in part to shifting interfaces, and in part to fragility at the layer of the template because it makes the logic of the application beholden to the structure of HTML.  But the most significant difficulty was simply in describing the interface between the template in a model in a way which an editor of the template could understand, and writing test failures in such a way that when the test failed, the template editor knew what happened.

Nevertheless it sounds like you've got some great use-cases here.  Can you provide links to any of this code, so we can see what your tests and code actually look like?  Perhaps your application could serve as a model for what the forward port of and/or replacement for patterns and data...s should include.

**Comment by twisted-trac** [comment]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)<a name="note_7"></a>|@bitglue commented|
|-|-|

Replying to [glyph](#note_6):
> Replying to [indigo](#note_4):
> 
> > Because renderers in Nevow can take parameters from both the view (patterns) and model (data, which I'm also missing) they are highly reusable.
> 
> I've actually found the opposite.  Because renderers can take parameters from the view and the model, determined by the template, the template becomes a functional component of the system which is hard to test. [...] [The template] makes the logic of the application beholden to the structure of HTML.

Decoupling the template, or at least funneling the interaction into well defined points, is the whole point of having patterns, as I understand it. If you don't have patterns, then you have to do one of three things:

1) create a separate template

2) embed view definitions (HTML) as stan in your code

3) manipulate the existing template in arbitrary ways, either using arbitrary tags from it as patterns, or inserting other content in arbitrary places

If you do #2 or #3, the interface to the template becomes complicated. #3 especially makes things hard to test, but this is exactly what most of the examples in the [template documentation](http://twistedmatrix.com/documents/current/web/howto/twisted-templates.html) do. [#11338](https://github.com/twisted/twisted/issues/11338) is inconvenient for reasons already discussed. In practice no one does it because #2 and #3 are so much easier.

If you say render methods can only manipulate the template by filling slots and cloning patterns, then testing and documentation are much easier. I can imagine something like z.i.Interface describing what render methods do, some facility that can statically analyze a template and find render method invocations and the associated Interface, and a verification method similar to z.i.v.verifyObject which then checks that the template (or the subset of it in scope to a particular render invocation) provides the necessary patterns and that each slot it has is something the renderer claims to fill. In this way your tests are essentially written for you once you've done the work of documentation.

I'm not entirely sure where "data" fits into this. There's no reason I can't write renderers like this:

```python
def render_assets(self, request, tag):
    return rend.sequence(request, tag, self.getAssets())
```

I like that it combines the "data" and "render" directives in the template into just one. That is good, since a specific "data" probably can't be rendered by an arbitrary renderer. That's one less thing that can be broken in the template. Sometimes I find it useful to use the "string" builtin renderer to verify the value of data, but this is a minor inconvenience. A print statement in the code would work almost as well.

What I don't like about it is that the render method is more difficult to test. I want to assert that the method:

- got the data

- rendered it with rend.sequence (which is complicated, and tested elsewhere)

which should be easy except it's not. I'll have to add a way to replace rend.sequence with something else to my class, and my test case will have to implement the replacement. Not impossible (or uncommon) by any means, but more difficult than the test I have to write with Nevow, which is none at all, because there is no render method. The render method now is essentially just glue, and I always dislike testing glue because glue interfaces with many things, each of which must be replaceable, and implemented in the test.

> Nevertheless it sounds like you've got some great use-cases here.  Can you provide links to any of this code, so we can see what your tests and code actually look like?  Perhaps your application could serve as a model for what the forward port of and/or replacement for patterns and data...s should include.

Unfortunately this code isn't something I'm allowed to share. I see no harm in including a few excerpts, though:

```
#!text/html
            <table class="form" n:render="clientInformation">
                <tr n:pattern="formInput">
                    <th>
                        <n:slot name="display">DISPLAY</n:slot><br/>
                        <span class="error"><n:slot name="error">ERROR</n:slot></span>
                    </th>
                    <td>
                        <input size="30">
                            <n:attr name="name"><n:slot name="name"/></n:attr>
                            <n:attr name="value"><n:slot name="value"/></n:attr>
                        </input>
                    </td>
                </tr>
                <tr n:pattern="immutableFormInput">
                    <th><n:slot name="display">DISPLAY</n:slot></th>
                    <td><n:slot name="value">VALUE</n:slot></td>
                </tr>
            </table>
```

This is one use case. render_clientInformation ends up invoking some code that generates forms, validates them, etc. It's too big to include, but this one bit probably says all you need to know about why I need patterns here:

```python
    def _buildClientForm(self, mutable):
        if mutable:
            pattern='formInput'
        else:
            pattern='immutableFormInput'
        self.clientForm = self.clientFormFactory(self.original.data, pattern)
```

It is in cases like this, where patterns provide options from which a rendering style will be selected, that having the patterns in the template is essential. I can strip out the nevow tags (a trivial task with XSLT) and run the template through an HTML validator to check the validity of my markup of when the client is mutable or immutable. I can also view the template in a browser when testing the graphical design of the page and see both cases without having to load application data such that I have both mutable and immutable clients.

rend.sequence is probably the more common case, but I don't use it for alternating row colors:

```
#!text/html
    <table n:data="assets" n:render="sequence">
        <tr n:pattern="header">
            <th>Type</th>
            <th>Asset #</th>
        </tr>
        <tr n:pattern="item" n:render="mapping">
            <td><n:slot name="type">TYPE</n:slot></td>
            <td><n:slot name="number">NUMBER</n:slot></td>
        </tr>
        <tr n:pattern="empty"><td colspan="2" style="text-align: center">No assets</td></tr>
    </table>
```

rather it's just a way to keep the view entirely in the template, and separate from the model or controller or whatever you'd like to call "not view". Consequently my render methods don't know if the assets are being rendered as a table, an unordered list, in a custom XML schema, etc. Not only is it more flexible, the render method and its test are simpler because it's easier to assert that a slot was filled with the correct value than assert the render method did whatever arbitrary thing was dictated by the view. 

Regarding tests, there are none of the templates, besides checks for XHTML validity that I run manually when I feel the need. Tests for data methods are not really noteworthy, and there are few render methods. Here's an example of one test (yeah probably I shouldn't look directly at d.result):

```python
    def test_render_ifEditable(self):
        d = self.clientPage.render_ifEditable(self.context, None)
        self.client._answerQueries(False) # emulates the completion of a query
        self.assertEqual(d.result, [])

        d = self.clientPage.render_ifEditable(self.context, None)
        self.client._answerQueries(True)
        self.assertEqual(d.result, self.context.tag)
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

Thanks for the comprehensive analysis of these issues. This is unfortunately going to take me quite some time to digest (but that shouldn't stop anyone else from working on this in the meanwhile).

**Comment by twisted-trac** [comment]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)<a name="note_9"></a>|@bitglue commented|
|-|-|

I think I've hit on a minimal description of the cases where I really miss patterns: rendering nested data. In the attached nestedData.py I use the case of a list of clients, each of which has 1 or more contacts as email addresses.

The problem is that one of two things need to happen:

1) The outer renderer (clients) needs to have nested for loops, and the inner loop needs to be able to clone some bit of template other than the 'tag' parameter

2) A separate renderer can render the inner data (contacts), but it needs some state from the outer loop (to know which client's contacts should be rendered)

Nevow provides either solution, through patterns or data, respectively. I concede only one is necessary or desirable. [[#11338](https://github.com/twisted/twisted/issues/11338)1338](https://github.com/twisted/twisted/issues/11338) seems like the more elegant option.

Solution #1 need not be any more complicated than an API on Tag that allows one to locate, and optionally remove, a child Tag given a specified identifier. I'm thinking an attribute t:id, and methods getChild(id) and popChild(id).

**Comment by twisted-trac** [comment]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)<a name="note_10"></a>|@bitglue commented|
|-|-|

Attached template-namespaces-5040.patch implements my proposed solution.

This is the result of a discussion on IRC between glyph and I. I think it's a very elegant solution, adding neither patterns nor data, but instead allowing one to change the namespace used for renderer lookups. It's somewhat similar in concept to functional closures. In the terms of the previous example, it lets one explicitly bind each "contact" renderer to a specific client.

Also attached (nestedDataWithNamespaces.py) is a rewrite of the original nested data example, solved with the new functionality.

One minor note: I was wondering if _StanLoader should be public or private. I suspect that if one were to create an instance giving it a structure that didn't contain only Tags but possibly other IRenderable instances, especially things that may be mutable or not reentrant, things might not work. I'm not sure of that, but I couldn't think of any specific reason one would need this API, so I opted to keep it private.

**Comment by twisted-trac** [comment]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)<a name="note_11"></a>|@bitglue commented|
|-|-|

Also, regarding documentation: if this patch is accepted, I promise to expand on the twisted.web.template howto on the use of Namespace, along with addressing the concerns in [#5286](https://github.com/twisted/twisted/issues/5286). I can't think of an example that isn't horribly contrived and doesn't build on the techniques there, which is why I didn't add documentation in this change.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @bitglue|
|-|-|

Thanks for your work on this, indigo.  I'm pretty sure I could be convinced that this is the right approach.  However, I would like to hear about why we shouldn't just make `_StanLoader` public and forget about `Namespace`.  Presenting this functionality as a _namespace_ seems a little bit weird to me.  `ClientNamespace` in the updated example is just a view object for a client model object.  Sure, it's also a namespace, but that's because pretty much anything in Python is a namespace.

Apart from that, `TestNamespace.test_flatten` needs a docstring, but maybe it will change significantly if `Namespace` doesn't end up being the approach taken here.

[#5286](https://github.com/twisted/twisted/issues/5286) is merged so hopefully the documentation can be expanded simultaneously with this feature addition now.

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)<a name="note_13"></a>|@bitglue removed owner|
|-|-|

Got rid of Namespace, made _StanLoader public (now TagLoader), and adjusted its .load() to return a list, as required by ITemplateLoader.

I don't really feel like extending the howto documentation anymore. Something about how its always broken with some obscure documentation toolkit I didn't know was in use but I never know until days after I submit the change for review. Make another ticket and I'll get to it some day.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [32834]) Branching to 'pattern-surrogate-5040'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [32835]) Apply template-namespaces-5040.2.patch

refs [#5040](https://github.com/twisted/twisted/issues/5040)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [32840]) Add an example of using TagLoader for subviews

refs [#5040](https://github.com/twisted/twisted/issues/5040)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks.  I think I'll take a crack at the documentation.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun removed owner|
|-|-|

See r32840.  [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/pattern-surrogate-5040)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph commented|
|-|-|

Replying to [indigo](#note_13):

> I don't really feel like extending the howto documentation anymore. Something about how its always broken with some obscure documentation toolkit I didn't know was in use but I never know until days after I submit the change for review. Make another ticket and I'll get to it some day.

Until that toolchain is fixed, please feel free to submit "broken" documentation and I'll fix it for you.  And then maybe punch the people who keep telling me that this new toolchain is what we need to "encourage" people to write more docs, a little.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_20"></a>|@glyph set owner to @exarkun|
|-|-|

The conflicts appear to be trivial, and the tests pass.  After our recent discussion I think this is the right approach.  Looks good to me, please land.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun commented|
|-|-|

(In [32906]) Branching to 'pattern-surrogate-5040-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun set status to `closed`|
|-|-|

(In [32911]) Merge pattern-surrogate-5040-2

Author: indigo, exarkun
Reviewer: exarkun, glyph
Fixes: [#5040](https://github.com/twisted/twisted/issues/5040)

Add documentation and a small amount of helper code to demonstrate
and simplify the web templating/rendering use case involving the
factoring of code responsible for rendering a number of different
objects.  This is a substitute for the pattern feature from Nevow
(on which `twisted.web.template` is heavily based): instead of
explicitly supporting named elements in the template, just make it
easier to re-use a whole sub-tree of the template with a different
Element, and re-use render directives to let the template interact
with Python code.
